File: python/124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    cur_max = float("-inf")
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        self.helper(root)
        return self.cur_max
    def helper(self, root):
        if not root:
            return 0
        left = self.helper(root.left)
        right = self.helper(root.right)
        left = 0 if not left else (left if left > 0 else 0)
        right = 0 if not right else (right if right > 0 else 0)
        self.cur_max = max(left + right + root.val, self.cur_max)
        return max(left, right) + root.val
    

File: cpp/101-150/124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
    	int res = INT_MIN;
    	maxtoroot(root, res);
    	return res;
    }
    int maxtoroot(TreeNode *root, int &res){
    	if(!root)
    		return 0;
    	int l = maxtoroot(root->left, res);
    	int r = maxtoroot(root->right, res);
    	if(l < 0)
    		l = 0;
    	if(r < 0)
    		r = 0;
    	res = max(res, l + r + root->val);
    	return root->val + max(l, r);
    }
};